### ISSUE TITLE
Add means to throw an error

### ISSUE THREAD

**Issue by cdavernas** [serverlessworkflow/specification#771]
## What would you like to be added?

Add an action that allows throwing an error

## Why is this needed?

Allows to define and throw workflow-specific errors, which is not possible as of now.

Coupled with the action's `condition` property, it could provide users a very powerfull - yet extremely simple - tool to perform flow and/or data validation logic.

## What is your proposal?

Create a new `throw` or `error` action type, that allows users to do the following:

```yaml
actions:
  - name: is-room-available
    functionRef:
      refName: check-if-room-is-available
      arguments:
        id: 123
    actionDataFilter:
      toStateData: .roomAvailable
  - name: throw-if-not-true
    condition: ${ .roomAvailable == false }
    errorRef:
      refName: room-not-available
```

If, as suggested in #770, we get rid of the top-level `errors` property, then it could instead look like:

```yaml
actions:
  - name: is-room-available
    functionRef:
      refName: check-if-room-is-available
      arguments:
        id: 123
    actionDataFilter:
      toStateData: .roomAvailable
  - name: throw-if-not-true
    condition: ${ .roomAvailable == false }
    throw:
      type: https://fake.hotel.com/errors/room-not-available
      status: 409
      title: Room Not Available
```

**Comment by fjtirado** [comment]
Including an error object has been requested by community
see 
https://kie.zulipchat.com/#narrow/stream/232676-kogito/topic/Exception.20Handling/near/360818735
and 
https://kie.zulipchat.com/#narrow/stream/232676-kogito/topic/Serverless.20Workflow.20error.20events

**Comment by fjtirado** [comment]
So basically we nee some $WORKFLOW.lastError or something like that so users can do stuff like the one requested in those zulip discussion

**Comment by fjtirado** [comment]
@cdavernas yes, they are basically asking how to know (from the state that is executed when the error happen) which was the error that causes the transition. This allow the same state handling different error types or extract info from the error itself

**Comment by cdavernas** [comment]
Do I have the go ahead to open a PR, then @ricardozanini @fjtirado @tsurdilo ?

**Comment by fjtirado** [comment]
Im going to play devils advocate, whats the diference between data condition on swith thats goes to an error branch and conditionally throw an error that goes to an error branch? ;)
Im not opposed to this, just a last time check.

**Comment by cdavernas** [comment]
Hmmm, that's a good question. I would be tempted to say there's no difference, and that in some case one way might be more convenient than the other, not sure. Do you think it poses any problem?

**Comment by tsurdilo** [comment]
I'm trying to understand this request. is this error thrown by workflow, so does it mean that workflow execution fails on 
input validation?
if this is the case, can this be done on input schema level, a flag that user can set to fail execution on data schema validaiton?

**Comment by cdavernas** [comment]
>I'm trying to understand this request. is this error thrown by workflow, so does it mean that workflow execution fails on
input validation?

No, not on input. That's the point. You throw based on a data evaluation, which is not predictable AOT, as it could very well be the output of a remote function

**Comment by tsurdilo** [comment]
I see, so its basically defining custom in-workflow-definition functions that contain some business logic. 
I wonder then if this could also be done on the event filter level?

**Comment by cdavernas** [comment]
> I wonder then if this could also be done on the event filter level?

It could, but that's not the point here. It's not really a filter, and is not necessarily related to events. 

Just see it as a way for user to choose to fault their workflow, for some reason. When combined with retries, this can be an absolutly fantastic feature, painless to ipmplement, and which has been requested a couple of times in the past.

**Comment by tsurdilo** [comment]
>  feature that allows users to throw an error based on runtime conditions somewhere in their flow

trying to understand ^^, its not somewhere its part of (from example) action chain
so what im trying to say is i think "why have to add an extra activity (or a set of) just to validate output of an action"? I think this is great idea but implementation-wise think could be added to maybe action definition itself (so no additional actions have to be added) or similar, maybe data filtering

**Comment by tsurdilo** [comment]
to add, what im thinking is more like:

          actions:
            - name: is-room-available
              functionRef:
                refName: check-if-room-is-available
                arguments:
                  id: 123
              actionDataFilter:
                toStateData: .roomAvailable
              <TIE TO THE EXCEPTION YOU HAVE HERE RATHER THAN SEPARATE ACTION>

**Comment by tsurdilo** [comment]
you could then also make these if/else checks with throw reusable components across multiple actions maybe

**Comment by tsurdilo** [comment]
> Not sure I understand what you mean by that

make these conditions reusable like we do for errors so 

     throw-if-not-true
    condition: ${ .roomAvailable == false }
    errorRef:
      refName: room-not-available


could be a reusable definition that you could chain on an action definitions rather than having to put each individual as action as in example

**Comment by cdavernas** [comment]
> could be a reusable definition that you could reference in an action definitions rather than having to put each individual as action as in example

Yeah, sure, reusability is indeed a big stress here, and we could outfactor the concept in a reusable component, but errors produced that way will be mostly flow/branch specific. Anyways, that's IMO another concern that I'm very happy with., but it still does not address the ability to explicitly and declaratively throw an error

**Comment by tsurdilo** [comment]
or something like:

         actions:
          - name: is-room-available
            functionRef:
              refName: check-if-room-is-available
              arguments:
                id: 123
            actionDataFilter:
              toStateData: .roomAvailable
            exceptionally:
                 - condition: "..."
                 - errorRef: "..."

**Comment by cdavernas** [comment]
That's just my opinion, but even though that could obviously do the trick, the last sample is a bit more verbose and a bit confusing that what's proposed... You know me and ubiquity, right 😆 !?

Instead of using yet another keyword that we would have to explain carefully (because not ubiquitous enough), I'd personally prefer that error action approach. Actions are IMO a bit bloated now, like kind of god-concept with a bit of everything: filtering, error handling, retries, ...
 
Again, just a personal opinion, but I'd rather go with my proposal, though modified to satisfy your perfectly rational concern regarding reusability.

Something like the first of my two examples:

```yaml
actions:
  - name: is-room-available
    functionRef:
      refName: check-if-room-is-available
      arguments:
        id: 123
    actionDataFilter:
      toStateData: .roomAvailable
  - name: throw-if-not-true
    condition: ${ .roomAvailable == false }
    errorRef:
      refName: room-not-available
```

**Comment by cdavernas** [comment]
Very good question. I do have dome ideas regarding that, as I'm sure you do, but yeah, with pleasure, let's go over that in meet 😉

**Comment by cdavernas** [comment]
Closed as resolved by 0.10, and therefore as part of #843
